lib/core/src/Cardano/Wallet/Address/Book.hs

```diff
@@ -168,26 +179,47 @@ instance ( key ~ SharedKey ) => AddressBookIso (Shared.SharedState n key)
         -- Trick: We keep the type, but we empty the discovered addresses
     data Discoveries (Shared.SharedState n key)
         = SharedDiscoveries
-            ( Map
-                (KeyFingerprint "payment" SharedKey)
-                (Index 'Soft 'ScriptK, AddressState)
-            )
+            (SharedAddressMap 'UtxoExternal key)
+            (SharedAddressMap 'UtxoInternal key)
 
     addressIso = iso from to
       where
         from st = case Shared.ready st of
-            Shared.Pending -> (SharedPrologue st, SharedDiscoveries Map.empty)
-            Shared.Active pool ->
-                let pool0 = AddressPool.clear pool
-                in  ( SharedPrologue st{ Shared.ready = Shared.Active pool0 }
-                    , SharedDiscoveries $ AddressPool.addresses pool
+            Shared.Pending ->
+                ( SharedPrologue st
+                , SharedDiscoveries (SharedAddressMap Map.empty) (SharedAddressMap Map.empty))
+            Shared.Active (Shared.SharedAddressPools extPool intPool _pending) ->
+                let extPool0 = clearShared extPool
+                    intPool0 = clearShared intPool
+                    pending0 = Shared.emptyPendingIxs
+                    pools0 = Shared.SharedAddressPools extPool0 intPool0 pending0
```

Comment:
```suggestion
            Shared.Active (Shared.SharedAddressPools extPool intPool pending) ->
                let extPool0 = clearShared extPool
                    intPool0 = clearShared intPool
                    pools0 = Shared.SharedAddressPools extPool0 intPool0 pending
```

😱 The pending indices need to be part of the Prologue; they are not part of the discoveries — the state of affairs is similar for the `SeqState`.

(And yes, this means that there could actually be a subtle problem with the order of pending addresses when a rollback happens — but this problem has always been there for the `SeqState`, this is not the right place for a fix.)

In any case, I do think the idea of grouping the external, internal and pending addresses into a single data type is very sound — 👍 to that.

Reply:
very true - done